gpfit/fit_constraintset.py

```diff
@@ -0,0 +1,160 @@
+            with NamedVariables("fit"):
+                dvars = VectorVariable(fitdata["d"], "u")
+                ivar = Variable("w")
+
+        self.dvars = dvars
+        self.ivar = ivar
+        self.rms_err = fitdata["rms_err"]
+        self.max_err = fitdata["max_err"]
+
+        monos = [fitdata["c%d" % k]*NomialArray(array(dvars).T**array(
+            [fitdata["e%d%d" % (k, i)] for i in
+             range(fitdata["d"])])).prod(NomialArray(dvars).ndim - 1) for k in
+                 range(fitdata["K"])]
+
+        if err_margin == "Max":
+            self.mfac = Variable("m_{fac-" + name + "-fit}",
+                                 1 + self.max_err, "-", "max error of fit")
+        elif err_margin == "RMS":
+            self.mfac = Variable("m_{fac-" + name + "-fit}",
+                                 1 + self.rms_err, "-", "RMS error of fit")
+        elif err_margin != None:
+            raise ValueError("Invalid name for err_margin: valid inputs Max, "
+                             "RMS")
+        else:
+            self.mfac = Variable("m_{fac-" + name + "-fit}", 1.0, "-",
+                                 "fit factor")
+
+        if fitdata["ftype"] == "ISMA":
+            # constraint of the form 1 >= c1*u1^exp1*u2^exp2*w^(-alpha) + ....
+            alpha = array([fitdata["a%d" % k] for k in
+                           range(fitdata["K"])])
+            lhs, rhs = 1, NomialArray(monos/(ivar/self.mfac)**alpha).sum(0)
+        elif fitdata["ftype"] == "SMA":
+            # constraint of the form w^alpha >= c1*u1^exp1 + c2*u2^exp2 +....
+            alpha = fitdata["a1"]
+            lhs, rhs = (ivar/self.mfac)**alpha, NomialArray(monos).sum(0)
+        elif fitdata["ftype"] == "MA":
+            # constraint of the form w >= c1*u1^exp1, w >= c2*u2^exp2, ....
+            lhs, rhs = (ivar/self.mfac), NomialArray(monos).T
+
+        if fitdata["K"] == 1:
+            # when possible, return an equality constraint
+            if hasattr(rhs, "shape"):
+                if rhs.ndim > 1:
+                    self.constraint = [(lh == rh) for lh, rh in zip(lhs, rhs)]
+                else:
+                    self.constraint = (lhs == rhs)
+            else:
+                self.constraint = (lhs == rhs)
+        else:
+            if hasattr(rhs, "shape"):
+                if rhs.ndim > 1:
+                    self.constraint = [(lh >= rh) for lh, rh in zip(lhs, rhs)]
+                else:
+                    self.constraint = (lhs >= rhs)
+            else:
+                self.constraint = (lhs >= rhs)
+
+        self.bounds = {}
+        for i, dvar in enumerate(self.dvars):
+            self.bounds[dvar] = [fitdata["lb%d" % i],
+                                 fitdata["ub%d" % i]]
+
+        ConstraintSet.__init__(self, [self.constraint])
+
+    def get_fitdata(self):
+        " return fit data "
+        return self.fitdata
+
+    def get_dataframe(self):
+        " return a pandas DataFrame of fit parameters "
```

Comment:
To further comment on this, from a designer perspective I want to pass in the constants and exponents to the fit constraint set versus alpha/A/B because I don't necessarily want to know/care about A/B/alpha.